### DISCUSSION TITLE
Stack de Software do Projeto

### DISCUSSION THREAD

**Discussion by MauricioMachadoFF** [UNB-PI2-Grupo3-BracoRobotico/Documentation#5] [General]
Olá pessoal! Queria abrir uma discussão para alinharmos a stack do projeto. Até agora pensei nas seguintes áreas que precisaremos alinhar.

## Áreas

- Backend
- Frontend
- Banco de Dados

Não detenho conhecimento de embarcado então não sei o que seria realizado nessa área. Se alguém puder comentar.

Outro ponto importante quanto a frontend será a plataforma que desenvolveremos para. Framework como Flutter permite nós criarmos aplicativos que podem ser distribuídos de diversas maneiras. Assim podemos adicionar uma continuidade do nosso projeto, como além da estação o lançamento de um aplicativo para coordenar o braço remotamente.

# Perguntas que devem ser respondidas

- Qual tecnologia iremos usar para desenvolvimento frontend?
  - Quais suas vantagens?
  - Quais suas desvantagens?
  - Qual o nível de conhecimento da equipe acerca dessa linguagem?
  - Existe alguma maneira de ajudarmos os membros quanto ao conhecimento? Como?
  - Como essa tecnologia se relaciona com demais áreas (embarcado, backend e afins). Atende todas as necessidades?

- Qual tecnologia iremos usar para desenvolvimento backend?
  - Quais suas vantagens?
  - Quais suas desvantagens?
  - Qual o nível de conhecimento da equipe acerca dessa linguagem?
  - Existe alguma maneira de ajudarmos os membros quanto ao conhecimento? Como?
  - Como essa tecnologia se relaciona com demais áreas (embarcado, banco de dados e afins). Atende todas as necessidades?

- Qual linguagem de banco de dados iremos usar?
  - Será um banco relacional ou não relacional? Por que?
  - Qual versão será usada?
  - Iremos utilizar docker para evitar problemas de diferentes versões operando o sistema?

- Como iremos embarcar nossa stack?
  - Quais possíveis placas temos para isso? Por que elas (viabilidade de preço, poder operacional, etc).
  - Insiram aqui mais perguntas sobre o sistema embarcado por favor.

**Comment by fmaachadoo** [comment]
Eu também não possuo familiaridade com sistemas embarcados, então fica difícil de opinar...

Sobre Front-end, nunca mexi com Flutter, mas acredito ser uma boa devido a versatilidade de deploy. Acho que vale a pena mencionar que sou mais familiarizado com React e Angular...

Sobre backend, pessoalmente falando, acredito que qualquer framework em python seja bem fácil de desenvolver...

Também vale a pena usar docker tanto no front-end, quanto no backend quanto no banco de dados. Acredito que isso é pelomenos o mínimo que podemos fazer pra tentar assegurar que todos os ambientes de desenvolvimento sejam iguais. Para evitar o problema de diferentes versões operando o sistema será necessário utilizar um gerenciador de dependências (no caso de python poderia ser um poetry ou pipenv) que defina as dependências e as versões das mesmas, que no caso, poderia estar dentro do container Docker.

Creio que também vale ressaltar que ainda não temos artefatos o suficiente que definam como o projeto vai funcionar, acredito que o produto seja definido antes da discussão sobre stack pois até o presente momento o que temos são esboços.

**Reply by fmaachadoo** [reply]
yarn funciona como gerenciador de dependências pro front-end?

**Reply by MauricioMachadoFF** [reply]
depende do framework a ser utilizado. O flutter por exemplo tem o arquivo puspec.yaml que faz toda essa gerência.

**Reply by MauricioMachadoFF** [reply]
acho que para frontend não é necessário o docker nesse caso. Podemos usar uma biblioteca que simplesmente aponta qual a versão de flutter que deve ser utilizada. Ela se chama fvm. Mas para outras linguagens pode ser sim também.

**Reply by fmaachadoo** [reply]
Acredito que seria bom se a gente usar Docker pra frontend também, visto que essas bibliotecas apenas mexem com gerências de dependências, e não com o ambiente de desenvolvimento como um todo. Usando Docker a gente consegue garantir o mesmo ambiente de desenvolvimento para todos os desenvolvedores, além do setup do projeto ficar mais simplificado.

**Comment by fmaachadoo** [comment]
Nós podemos utilizar serviços na cloud? ou todos os softwares que utilizaremos terá que estar dentro do hardware do projeto?

**Reply by fmaachadoo** [reply]
Por exemplo, o backend, estaria rodando na cloud? ou em um raspberry ou coisa do tipo?

**Reply by MauricioMachadoFF** [reply]
Podemos ter ambos. Por exemplo: a gerência de estoque é cloud.

**Comment by fmaachadoo** [comment]
Acredito que podemos usar Apache Kafka como RPC (remote procedure call) para ajudar na comunicação do backend, frontend e do sistema embarcado. 
Acredito que seria útil tanto ao criar o pedido do cliente, quanto pro braço robótico iniciar a ação de pegar o remédio, tanto pra informar o cliente em qual estado do sistema que o procedimento se encontra.

**Comment by PedroMoraes39** [comment]
Se formos usar a raspberry acho que conseguimos rodar a parte do back nela mesma, o que teríamos que nos preocupar e como fazer a comunicação do aplicativo, como seria feito por Bluetooth ou seria via cabo. 
Esse e uma questão par alinhar com a parte da eletrônica

**Reply by fmaachadoo** [reply]
é viável da conexão acontecer via wifi?

**Reply by fmaachadoo** [reply]
não necessariamente via wifi, mas usando HTTP ou outro protocolo de comunicação pela internet

**Reply by PedroMoraes39** [reply]
acho que sim, mas ai teria que ter uma placa wifi no braço, acho que seria o mesmo principio do bluetooth. No final temos que ver o que vai ser mais fácil de fazer o reconhecimento com o back e tmb qual tem o menor custo

**Reply by fmaachadoo** [reply]
Eu dei uma idéia de fazer a comunicação usando um sistema pub/sub, se for viável colocar uma placa wifi eu acho que rola

**Reply by PedroMoraes39** [reply]
Seria uma boa, mas teríamos que conversar com o pessoal de eletrônica para ver como vai ficar essa parte

**Comment by PedroMoraes39** [comment]
Sobre a questão dos bancos eu acho so bancos relacionais mais fáceis de utilizar, poderíamos utilizar python para fazer a leitura juntamente como sqlalchemy e o pandas (pelo menos foram as bibliotecas que ja utilizei)

**Comment by MauricioMachadoFF** [comment]
Hoje definimos o tech stack do projeto durante reunião na quarta feira em aula conforme acordamos.

A stack ficou da seguint forma:

* Frontend: Flutter
* Backend: Python
* Banco de dados: PostgreSQL

# Por que essa escolha

Flutter: Versatilidade de dispositivos para uso, podemos adicionar como um próximos passos o deploy de um app nas lojas da google e apple. Outro ponto é sobre conseguirmos compilar um app direto para o linux o que facilita ao embarcarmos o OS na RASP 4.

Python: Possui bibliotecas que facilitam a integração para com o sistema embarcado. Além disso, grande parte de equipe tem experiência com essa linguagem.

Banco de Dados: Relacional devido as necessidade do trabalho.
